modules/assim.sequential/R/load_data_paleon_sda.R

```diff
@@ -0,0 +1,148 @@
+  settings$host$name <- "localhost"
+  
+  site <- PEcAn.DB::query.site(settings$run$site$id, bety$con)
+  format_id <- settings$state.data.assimilation$data$format_id
+  
+  input.list <- list()
+  input.id <- list()
+  obvs <- list()
+  
+  var.names <- unlist(sapply(settings$state.data.assimilation$state.variable, 
+                             function(x) {
+                               x$variable.name
+                             }, 
+                             USE.NAMES = FALSE), 
+                      use.names = FALSE)
+  
+  start_date <- settings$state.data.assimilation$start.date
+  end_date   <- settings$state.data.assimilation$end.date
+  
+  obs.mean <- list()
+  obs.cov <- list()
+  
+  start.time <- format(lubridate::ymd(start_date),settings$state.data.assimilation$forecast.time.step)
+  end.time <- format(lubridate::ymd(end_date),settings$state.data.assimilation$forecast.time.step)
+  obs.times <- start.time:end.time
+  
+  
+  for(i in seq_along(format_id)){
+    input.list[[i]] <- db.query(paste("SELECT * FROM inputs WHERE site_id =",site$id ,"  AND format_id = ",format_id[[i]]), bety$con)
+    input.id[[i]] <- input.list[[i]]$id
+    
+    data.path <- PEcAn.DB::query.file.path(input.id[[i]], settings$host$name, bety$con)
+    format_full <- format <- PEcAn.DB::query.format.vars(input.id = input.id[[i]], bety, format.id = NA, var.ids=NA)
+    
+    format$na.strings <- 'NA'
+    
+    # ---- LOAD INPUT DATA ---- #
+    time.row <- format$time.row
+    vars.used.index <- setdiff(seq_along(format$vars$variable_id), format$time.row)
+    
+    print(paste('Using PEcAn.benchmark::load_data.R on format_id',format_id[[i]],'-- may take a few minutes'))
+    obvs[[i]] <- PEcAn.benchmark::load_data(data.path, format, start_year = lubridate::year(start_date), end_year = lubridate::year(end_date), site, vars.used.index, time.row)
+    
+    dataset <- obvs[[i]]
+    variable <- var.names
+    
+    ### Tree Ring Data Product
+    if(format_id[[i]] == '1000000040'){
+      obvs[[i]] <- obvs[[i]][obvs[[i]]$model_type=='Model RW + Census',]
+      obvs[[i]]$AbvGrndWood <- obvs[[i]]$AbvGrndWood * .48
+      obvs[[i]]$NPP_1_C <- obvs[[i]]$NPP_1_C * .48
+      arguments <- list(.(year, MCMC_iteration, site_id), .(variable))
+      arguments2 <- list(.(year), .(variable))
+      arguments3 <- list(.(MCMC_iteration), .(variable), .(year))
+    }else{
+      print('ERROR: This data format has not been added to this function (ツ)_/¯ ')
+      stop()
+    }
+    
+    ### Map species to model specific PFTs
+    if(any(var.names == 'AGB.pft')){
+      spp_id <- match_species_id(unique(dataset$species_id),format_name = 'usda',bety)
+      spp_id <- spp_id[spp_id$input_code!='HAVI4',]
+      pft_mat <- match_pft(spp_id$bety_species_id, settings$pfts, con = bety$con)
+      
+      x <- paste0('AGB.pft.', pft_mat$pft)
+      names(x) <- spp_id$input_code
```

Comment:
I'm not following what's going on here -- x appears to have length pft_mat, which will <= to the length of spp_id, so unless there's a 1:1 match between species and PFT this will fail.  I think the logic here, and below, is too LINKAGES specific. More generally you need to sum species-level AGB's to the PFT level, such as with a tapply, with the 1:1 mapping being a special case.